Release Skin node (#3566)
cyberbotics/webots: docs/guide/humans.md

# Humans

> This chapter gives an overview of different humans simulation provided with the Webots package.

Reviewer: ```suggestion
This chapter gives an overview of different human simulations provided with the Webots package.
```

The example worlds can be tested easily; the ".wbt" files are located in various "worlds" directories of the "[WEBOTS\_HOME/projects/humans]({{ url.github_tree }}/projects/humans)" directory and can be directly opened from Webots using the `Open Sample World` item in `File` menu.
The controller code is located in the corresponding "controllers" directory.

## [skin\_animated\_humans]({{ url.github_tree }}/projects/humans/skin_animated_humans/worlds/skin_animated_humans.wbt) Sample

![skin_animated_humans.png](images/humans/skin_animated_humans.thumbnail.jpg) This example shows how to use the [Skin](../reference/skin.md) node to animate human models and play [BVH](https://en.wikipedia.org/wiki/Biovision_Hierarchy) animations.
The [FBX](https://en.wikipedia.org/wiki/FBX) models have been generated using [MakeHuman](http://makehumancommunity.org/) and BVH files are taken from open source databases.

> The [Skin PROTO models]({{ url.github_tree }}/projects/humans/skin_animated_humans/protos/) contain the definition of the appearance declared in the in the FBX files.

Reviewer: ```suggestion
The [Skin PROTO models]({{ url.github_tree }}/projects/humans/skin_animated_humans/protos/) contain the definition of the appearance declared in the FBX files.
```

> The "[bvh\_animation]({{ url.github_tree }}/projects/humans/skin_animated_humans/controllers/bvh_animation.c)" C controller loads the BVH file, associates it with the [Skin](../reference/skin.md) model and plays the BVH animation by applying the poses to the [Skin](../reference/skin.md) bones.

Reviewer: ```suggestion
The "[bvh\_animation]({{ url.github_tree }}/projects/humans/skin_animated_humans/controllers/bvh_animation/bvh_animation.c)" C controller loads the BVH file, associates it with the [Skin](../reference/skin.md) model and plays the BVH animation by applying the poses to the [Skin](../reference/skin.md) bones.
```

  SFRotation  rotation    0 1 0 0
  SFVec3f     scale       1 1 1
  SFString    name        "skin"
  SFString    model       "Sandra"
  SFBool      castShadows FALSE
}
```

> **File location**: "[WEBOTS\_HOME/projects/humans/skin\_animated\_humans/protos/CharacterSkin.proto]({{ url.github_tree }}/projects/humans/skin_animated_humans/protos/CharacterSkin.proto)"

#### CharacterSkin Field Summary

- `model` : Defines which character should be used.
This field accepts the following values: "Anthony", "Robert", "Sandra", and "Sophia".

### [bvh\_util]({{ url.github_tree }}/projects/humans/skin_animated_humans/libraries/bvh_util/) Library
To load and play the BVH file a C utility library is also provided.
The BVH file defines a skeleton in the form of hierarchy of bones, and a time series of angles for each joint in the skeleton.
Note that this skeleton is not necessarily the same as the skeleton associated with a [Skin](../reference/skin.md) node, since they come from different sources.
For this reason, this library also provides functions to adapt and re-target the BVH motion data to the [Skin](../reference/skin.md) model.

---

#### `wbu_bvh_read_file`

> #### `wbu_bvh_cleanup`

Reviewer: The description of the function are built in another way than in other .md files (like https://cyberbotics.com/doc/reference/accelerometer), with the italic text before the tab-component for example.
I don't know if its matter.

Author: Yes, I'm aware of this.
But the documentation of other functions in the reference manual has a complete different structure.
I checked different option and for me this one is the one resulting in a more clear structure.

Returns TRUE on success.

The function `wbu_bvh_goto_frame` reads a specific frame, whose index is the argument `frame_number`.
Returns TRUE on success.
Note that if the argument `frame_number` is greater than the number of frames in the file, it leads to an error, and the function returns FALSE.

The function `wbu_bvh_reset` jumps to the first frame.
Returns TRUE on success.

---

#### `wbu_bvh_get_joint_rotation`
#### `wbu_bvh_get_root_translation`

*get the joint rotation for a specific joint and the BVH object translation*

%tab-component "language"

%tab "C"

```c
double *wbu_bvh_get_joint_rotation(WbBvhDataRef ref, int joint_index);
double *wbu_bvh_get_root_translation(WbBvhDataRef ref);
```

%tab-end

%end

The `wbu_bvh_get_joint_rotation` function returns the rotation of the joint specified by `joint_index`.
The rotation is returned as an array of double.

The `wbu_bvh_get_root_translation` function returns the translation of the BVH object.
The translation is returned as an array of double.
This is typically used to set the translation of the [Skin](../reference/skin.md) node that is being animated by this BVH file.

> The values can wither be used directly, or scaled using the `wbu_bvh_set_scale` function.

Reviewer: ```suggestion
The values can either be used directly, or scaled using the `wbu_bvh_set_scale` function.
```

## [C3D]({{ url.github_tree }}/projects/humans/c3d/worlds/c3d_viewer.wbt) Sample

![c3d_viewer.png](images/humans/c3d_viewer.thumbnail.jpg) This example provides a basic viewer for [C3D](https://c3d.org) files containing biomechanical information and motion capture data.

> The "[c3d\_viewer.py]({{ url.github_tree }}/projects/humans/c3d/controllers/c3d_viewer.py)" Python controller loads the C3D file, creates a solid [Sphere](../reference/sphere.md), and plays the motion by updating the position of the markers.

Reviewer: ```suggestion
The "[c3d\_viewer.py]({{ url.github_tree }}/projects/humans/c3d/controllers/c3d_viewer/c3d_viewer.py)" Python controller loads the C3D file, creates a solid [Sphere](../reference/sphere.md), and plays the motion by updating the position of the markers.
```

> Additionally, "[c3d\_viewer.py]({{ url.github_tree }}/projects/humans/c3d/controllers/c3d_viewer.py)" also initializes the "[c3d\_viewer\_window]({{ url.github_tree }}/projects/humans/c3d/plugins/robot_windows/c3d_viewer_window)" robot window from which it is possible to set the transparency of the body representation and the playback speed, to change the color of the markers, and to visualize the graphs of the data provided in the C3D file.

Reviewer: ```suggestion
Additionally, "[c3d\_viewer.py]({{ url.github_tree }}/projects/humans/c3d/controllers/c3d_viewer/c3d_viewer.py)" also initializes the "[c3d\_viewer\_window]({{ url.github_tree }}/projects/humans/c3d/plugins/robot_windows/c3d_viewer_window)" robot window from which it is possible to set the transparency of the body representation and the playback speed, to change the color of the markers, and to visualize the graphs of the data provided in the C3D file.
```

  SFRotation  rotation                         0 1 0 0
  SFString    c3dFile                          ""
  SFFloat     playbackSpeed                    1.0
  MFNode      markers                          []
  SFVec3f     leftGroundReactionForceOffset    0.28 0 0.28
  SFVec3f     rightGroundReactionForceOffset   -0.21 0 0.22
  SFFloat     bodyTransparency                 1
  SFFloat     bodyHeight                       -1
  SFFloat     bodyOffset                       0.25
  SFString    window                           "c3d_viewer_window"
  SFString    name                             "c3d_viewer"
}
```

> **File location**: "[WEBOTS\_HOME/projects/humans/c3d/protos/C3dViewer.proto]({{ url.github_tree }}/projects/humans/c3d/protos/C3dViewer.proto)"

#### C3dViewer Field Summary

- `c3dFile` : Defines which file should be displayed.
- `playbackSpeed` : Defines the speed at which should be played the C3D file.
- `markers`: The markers will be automatically inserted in this field.
- `leftGroundReactionForceOffset`: Defines the offset of the left Ground Reaction Force position.
- `rightGroundReactionForceOffset`: Defines the offset of the right Ground Reaction Force position.

> - `bodyTransparency`: Defines the transparency of the body representation (transparency to 1 disable body representation).

Author: This is a copy from the field description in `C3DViewer.proto`.
I agree that the sentence is not very clear, but the original sentence is more correct than the suggestion.
With `to+` prefix verb should not be conjugated.

> - `bodyHeight`: Defines the height of the body representation, if set to '<0' the height defined in the C3D file is used (or 1.83m is not set in the C3D file).

Reviewer: ```suggestion
- `bodyHeight`: Defines the height of the body representation, if set to '<0' the height defined in the C3D file is used (or 1.83m if not set in the C3D file).
```

Author: I will fix the PROTO as well.

> - `bodyHeight`: Defines the height of the body representation, if set to '<0' the height defined in the C3D file is used (or 1.83m if not set in the C3D file).

Author: I will also fix the C3DViewer PROTO file accordingly.

- `bodyOffset`: Defines the vertical offset of the body representation compared to the 'CenterOfMass' marker.

## [Pedestrian]({{ url.github_tree }}/projects/humans/pedestrian/worlds/pedestrian.wbt) Sample

![pedestrian_world.png](images/humans/pedestrian_world.thumbnail.jpg) This example shows the simulation of simple walking rigid human models.

> The walking gait is hard-coded in the "[pedestrian]({{ url.github_tree }}/projects/humans/pedestrian/controllers/pedestrian.py)" Python controller source code, but the trajectory, the speed and the time step can be passed as controller argument:

Reviewer: ```suggestion
The walking gait is hard-coded in the "[pedestrian]({{ url.github_tree }}/projects/humans/pedestrian/controllers/pedestrian/pedestrian.py)" Python controller source code, but the trajectory, the speed and the time step can be passed as controller argument:
```
